packages/ui-components/src/components/PopupMenu/PopupMenu.component.tsx

```diff
@@ -0,0 +1,394 @@
+type HeadlessMenuItemsProps = React.ComponentPropsWithRef<React.ElementType>
+// Extend the extracted props instead:
+export interface PopupMenuMenuProps extends HeadlessMenuItemsProps {
+  as?: React.ElementType // Allow customizing the element type just as headless ui does
+  className?: string
+  children?: React.ReactNode
+  key?: React.Key
+}
+
+export interface PopupMenuItemProps extends React.ComponentProps<typeof HeadlessMenu.Item> {
+  as?: React.ElementType // Allow customising the element type to allow for rendering items as anchor elements, just as headless ui does
+  className?: string
+  disabled?: boolean
+  href?: string // accept a href when rendering a link, whether or not to pass it to HUI item will behandled in the component
+  icon?: keyof typeof KnownIconsEnum
+  label?: string
+  rel?: string // accept rel for anchors, handle in component
+  target?: string // accepot target for links, handle in component
+}
+
+export interface PopupMenuSectionProps {
+  children?: React.ReactNode
+  className?: string
+}
+
+export interface PopupMenuSectionHeadingProps {
+  children?: React.ReactNode
+  className?: string
+  label?: string
+}
+
+export interface PopupMenuSectionSeparatorProps {
+  className?: string
+}
+
+// ----- Define PopupMenu context and hook -----
+
+// Context:
+const PopupMenuContext = createContext<PopupMenuContextType | null>(null)
+
+// Hook to use the context:
+export const usePopupMenuContext = () => {
+  const context = useContext(PopupMenuContext)
+  if (!context) {
+    throw new Error(
+      "The usePopupMenuContext hook must be used inside within a PopupMenu that provides the PopupMenuContext."
+    )
+  }
+  return context
+}
+
+// Export the context for advanced use cases:
+export { PopupMenuContext }
+
+// ----- Component Definitions -----
+
+// POPUP MENU
+
+/** A Popup Menu component that wraps Headless UI Menu. The Menu will be rendered into a Juno Portal, so using Juno's PortalProvider (which is already included when using Juno's AppShell) is mandatory. */
+const PopupMenu: React.FC<PopupMenuProps> & {
+  Toggle: React.FC<PopupMenuToggleProps>
+  Menu: typeof PopupMenuMenu
+  Item: React.FC<PopupMenuItemProps>
+  Section: React.FC<PopupMenuSectionProps>
+  SectionHeading: React.FC<PopupMenuSectionHeadingProps>
+  SectionSeparator: React.FC<PopupMenuSectionSeparatorProps>
+} = ({
+  children = null,
+  className = "",
+  disabled = false,
+  icon = "moreVert",
+  menuSize = "normal",
+  onClose = undefined,
+  onOpen = undefined,
+  ...props
+}) => {
+  // Create a state to track headless-ui's internal open state (Note that we can track the internal state and can let headless ui update our tracking state, but we can not set it, hence there is no `open` prop on our component):
```

Comment:
so many comments here in PR in general, I do not see personally any values in them,  looks like generated content for me. Please, if it is possible remove them.

Reply:
I disagree. Especially for complex components like this that has to deal with/work around idiosyncrasies of the underlying headless library I'd rather err on the side of slightly too many comments. They help in months' or years' time when we have to try to understand why we do things a certain way.

Reply:
I agree 100% with Esther here. This is a pretty complex component, and by wrapping Headless Menu we have to apply some approaches that are not obvious and far from being self explanatory. Not sure what makes you think these are generated though?